### PR TITLE
Add tests for knative trigger in serverless-controller-manager integration tests

### PR DESCRIPTION
**Pull request by aerfio**
**Description**

Changes proposed in this pull request:

- test integration with brokers and triggers -> we ping broker to generate event, which we then catch in function, increasing some counter
- fix few bugs here and there
- use long polling in order to be sure that we wait long enough for answer

**Related issue(s)**

#8259

### CODE REVIEW

File: tests/function-controller/pkg/apirule/apirule.go [hunk 1]
```diff
@@ -85,3 +91,71 @@ func (a *APIRule) Delete() error {
...
+func (a *APIRule) WaitForStatusRunning() error {
+	apirule, err := a.get()
+	if err != nil {
+		return err
+	}
+
+	if isStateReady(apirule.Status) {
+		a.log.Logf("%s is ready:\n%v", a.name, apirule)
+		return nil
+	}
```
Comment (aerfio) [inline tests/function-controller/pkg/apirule/apirule.go:119]: this prevents subtle bug in which apirule is already ready, so you get no events in `watchtools.Until` function, which just hangs for infinity

File: tests/function-controller/pkg/broker/broker.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+func (b *Broker) WaitForStatusRunning() error {
+	broker, err := b.get()
+	if err != nil {
+		return err
+	}
+
+	if broker.Status.IsReady() {
+		b.log.Logf("%s is ready:\n%v", b.name, broker)
+		return nil
+	}
```
Comment (aerfio) [inline tests/function-controller/pkg/broker/broker.go:78]: this prevents subtle bug in which broker is already ready, so you get no events in `watchtools.Until` function, which just hangs for infinity

File: tests/function-controller/pkg/namespace/namespace.go [hunk 3]
```diff
@@ -1,36 +1,77 @@
...
+	}
+
+	backoff := wait.Backoff{
+		Duration: 500 * time.Millisecond,
+		Factor:   2,
+		Jitter:   0.1,
+		Steps:    4,
+	}
+
+	err := k8sretry.OnError(backoff, func(err error) bool {
```
Comment (aerfio) [inline tests/function-controller/pkg/namespace/namespace.go:57]: note to self - revert this change

File: tests/function-controller/Dockerfile [hunk 4]
```diff
@@ -1,4 +1,4 @@
-FROM golang:1.13.5-alpine3.10 as builder
+FROM golang:1.14-alpine3.11 as builder
```
Comment (aerfio) [inline tests/function-controller/Dockerfile:1]: golang 1.14 let's you stream test logs, so it's nice for DX

File: tests/function-controller/testsuite/testsuite.go [hunk 5]
```diff
@@ -70,30 +90,78 @@ func New(restConfig *rest.Config, cfg Config, t *testing.T, g *gomega.GomegaWith
...
+	functionDetails := t.getFunction(helloWorld)
+	fnResourceVersion, err := t.function.Create(functionDetails)
 	failOnError(t.g, err)
 
 	t.t.Log("Waiting for function to have ready phase...")
-	err = t.function.WaitForStatusRunning(resourceVersion)
+	err = t.function.WaitForStatusRunning(fnResourceVersion)
+	failOnError(t.g, err)
+
+	t.t.Log("Waiting for broker to have ready phase...")
```
Comment (magicmatatjahu) [inline tests/function-controller/testsuite/testsuite.go:106]: Should we wait first for broker, then for function?

File: tests/function-controller/main_test.go [hunk 6]
```diff
@@ -16,12 +19,14 @@ type config struct {
...
 func TestFunctionController(t *testing.T) {
+	rand.Seed(time.Now().UnixNano())
+
 	g := gomega.NewGomegaWithT(t)
 
 	cfg, err := loadConfig("APP")
 	failOnError(g, err)
 
-	restConfig, err := newRestClientConfig(cfg.KubeconfigPath)
+	restConfig := controllerruntime.GetConfigOrDie()
```
[the remark this answers is not in the document]
Comment (aerfio) [reply tests/function-controller/main_test.go:29]: then he is still able, just have this kubeconfig as your active one and it'll work

[the remark this answers is not in the document]
Comment (aerfio) [reply tests/function-controller/main_test.go:29]: in functionality nothing changed, user has even less things to do to have it work, but maybe I should document it

[the remark this answers is not in the document]
Comment (aerfio) [reply tests/function-controller/main_test.go:29]: but then again -> you can't run those tests on your own now, because there are internal calls to various services, so running it outside the cluster doesn't make sense

File: tests/function-controller/pkg/apirule/apirule.go [hunk 7]
```diff
@@ -85,3 +91,73 @@ func (a *APIRule) Delete() error {
...
+			a.log.Logf("Apirule %s is ready:\n%v", name, u)
+			return true, nil
+		}
+
+		a.log.Logf("Apirule %s is not ready", name)
+		return false, nil
+	}
+}
+
+func isStateReady(a apiruleTypes.APIRuleStatus) bool {
```
[the remark this answers is not in the document]
Comment (aerfio) [reply tests/function-controller/pkg/apirule/apirule.go:161]: If I wanted to do func (a *APIRule) then I'd have to do a get for function here, because APIRule is
```
type APIRule struct {
	resCli      *resource.Resource
	name        string
	namespace   string
	waitTimeout time.Duration
	log         shared.Logger
}
```

and I can't create methods for the types I do not own

File: tests/function-controller/main_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (magicmatatjahu) [reply tests/function-controller/main_test.go:29]: Not outside - inside, but with another cluster :P

File: tests/function-controller/pkg/apirule/apirule.go [hunk 8]
```diff
@@ -85,3 +91,73 @@ func (a *APIRule) Delete() error {
...
+	apirule, err := a.get()
+	if err != nil {
+		return err
+	}
+
+	if isStateReady(apirule.Status) {
+		a.log.Logf("%s is ready", a.name)
+		return nil
+	}
+	a.log.Logf("ApiRule %s is not ready", a.name)
```
[the remark this answers is not in the document]
Comment (aerfio) [reply tests/function-controller/pkg/apirule/apirule.go:120]: I'll rewrite it a bit 👍

File: tests/function-controller/pkg/apirule/apirule.go [hunk 9]
```diff
@@ -85,3 +91,73 @@ func (a *APIRule) Delete() error {
...
+}
+
+func (a *APIRule) WaitForStatusRunning() error {
+	apirule, err := a.get()
+	if err != nil {
+		return err
+	}
+
+	if isStateReady(apirule.Status) {
+		a.log.Logf("%s is ready", a.name)
```
[the remark this answers is not in the document]
Comment (aerfio) [reply tests/function-controller/pkg/apirule/apirule.go:117]: read https://github.com/kyma-project/kyma/pull/8312#discussion_r417828393

File: tests/function-controller/pkg/apirule/apirule.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (magicmatatjahu) [reply tests/function-controller/pkg/apirule/apirule.go:161]: I mean about `func (a *APIRule) isStateReady(a apiruleTypes.APIRuleStatus) bool {}`, not to changing args of function.

File: tests/function-controller/pkg/resource/resource.go [hunk 10]
```diff
@@ -122,3 +135,37 @@ func (r *Resource) Delete(name string, timeout time.Duration) error {
 	}
 	return nil
 }
+
+func (r *Resource) Update(res interface{}) (*unstructured.Unstructured, error) {
```
[the remark this answers is not in the document]
Comment (aerfio) [reply tests/function-controller/pkg/resource/resource.go:139]: That line is incorrect, you should rely on "k8s.io/apimachinery/pkg/api/errors".IsNotFound 

and it's modifying its argument 😢 

but I'll create service for converting things from unstructured to concrete type

File: tests/function-controller/testsuite/testsuite.go (hunk 5 above)
Comment (aerfio) [reply tests/function-controller/testsuite/testsuite.go:106]: nah, because knative creates broker along with properly labeled namespace, so if I wait for broker immediately after namespace creation I'd have to handle the case where the broker is not yet created, and this way it's 100% sure that broker is there, but I just check its status

Comment (aerfio) [reply tests/function-controller/testsuite/testsuite.go:106]: it's either this order (which, in truth, doesn't matter - trigger just needs to be created after broker, even better when created after broker is ready, which I already do) or handling whole new case - for me the choice is damn obvious

File: tests/function-controller/pkg/shared/shared.go [hunk 11]
```diff
@@ -1,5 +1,25 @@
...
+
+var (
+	ErrInvalidDataType = errors.New("invalid data type")
+)
+
+type Container struct {
+	DynamicCli  dynamic.Interface
+	Namespace   string
+	WaitTimeout time.Duration
+	Kind        string
```
Comment (magicmatatjahu) [inline tests/function-controller/pkg/shared/shared.go:22]: Kind? I don't see any usage of this field.

Comment (aerfio) [reply tests/function-controller/pkg/shared/shared.go:22]: I'll take care of this in next PR

File: tests/function-controller/README.md [hunk 12]
```diff
@@ -60,15 +60,16 @@ Use the following environment variables to configure the application:
...
 | **APP_TEST_INGRESS_HOST**               | No       | `kyma.local`               | The Ingress host address                                                                                                                              |
 | **APP_TEST_DOMAIN_PORT**                | No       | `80`                       | The port of the Service exposed by the API Rule in a given domain                                                                                     |
 | **APP_TEST_INSECURE_SKIP_VERIFY**       | No       | `true`                     | The flag that controls whether tests use verification of the server's certificate and the host name to reach the Function                                       |
+| **APP_TEST_VERBOSE**                    | No       | `true`                     | The value that controls whether tests log resources that are subject to change | 
+| **APP_TEST_MAX_POLLING_TIME**           | No       | `5m`                       | The maximum period of time that we wait for function to reconfigure after update  |
```
Comment (kazydek) [inline tests/function-controller/README.md:72]:
```suggestion
| **APP_TEST_MAX_POLLING_TIME**           | No       | `5m`                       | The maximum period of time in which the Function can reconfigure after an update  | 
```

Comment (kazydek) [reply tests/function-controller/README.md:72]: of sth other than "we" ;) like "the test waits"?

Comment (aerfio) [inline tests/function-controller/README.md:72]:
```suggestion
| **APP_TEST_MAX_POLLING_TIME**           | No       | `5m`                       | The maximum period of time in which the Function must reconfigure after an update  | 
```
